### DISCUSSION TITLE
Custom Styles

### DISCUSSION THREAD

**Discussion by gili-labs** [nalexn/ViewInspector#84] [Ideas]
I am unit testing a custom control that separates the concerns of appearance and behavior using styles. I'm looking at the code in ViewInspector that supports styles, such as ButtonStyle, with the goal of trying to determine a way to support custom styles. Does anyone have any ideas?

**Comment by gili-labs** [comment]
What is the difference between ViewType.ParentView and ViewType.ClassifiedView?

On the surface, they appear nearly equal in all regards, except ViewType.ParentView cannot conform to MultipleViewContent. However, I'm left wondering why?

Is there some subtle reasoning I'm missing.

**Reply by nalexn** [reply]
Hey @gili-labs , you're right, there is almost no difference between `ParentView` and `ClassifiedView` from the code, however, I separated these instances to emphasize their different semantics: referencing type-erased parent and child views respectively. I recall at some point I had to distinguish these two cases for proper `pathToRoot` calculation, apparently, that code was removed after a refactoring, but structs remained separated.

As for the original question you had, I have several examples of testing the custom styles in the [guide](https://github.com/nalexn/ViewInspector/blob/master/guide.md#inspection-guide), I'll be able to help if you ask a more specific question about the custom style you want to test.

**Reply by gili-labs** [reply]
I have read the guide, and it has provided me with some ideas. However, I'm not trying to test a custom style for a Swift UI view (e.g., Button or Label), rather I'm trying to test a custom style for a custom view.

Consider the following example:

```Swift
struct Content: View {
	var body: some View {
		ZStack {
			...
			MyControl()
				.environment(\.myControlStyle, CustomControlStyle)
			...
		}
	}
}

struct CustomControlStyle: MyControlStyle {
	public makeBody(configuration: MyControlStyleConfiguration) -> some View {
		CustomControl(configuration: configuration)
	}

	struct CustomControl: View {
		let configuration: MyControlStyleConfiguration

		var body: some View {
			...
 		}
	}
}
```

I want to test CustomControlStyle in the same way the ViewInspector framework does for styles defined by SwiftUI. For example:

```Swift
extension CustomControlStyle {
	func inspect(isDragging: Bool) throws -> InspectableView<ViewType.ClassifiedView> {
		let configuration = MyControlStyleConfiguration(isDragging: isDragging)
		let view = makeBody(configuration: configuration).inspect()
		return .init(view.content, parent: nil, index: nil)  // CANNOT DO THIS
	}
}

func testCustomControlStyleIsDragging() {
	let style = CustomControlStyle()
	let inspectedStyle = try style.inspect(isDragging: true)
	...
}
```

This follows what ViewInspector does for SwiftUI styles. For example:

```Swift
public extension ToggleStyle {
    func inspect(isOn: Bool) throws -> InspectableView<ViewType.ClassifiedView> {
        let config = ToggleStyleConfiguration(isOn: isOn)
        let view = try makeBody(configuration: config).inspect()
        return try .init(view.content, parent: nil, index: nil)
    }
}
```

Unfortunately, I can't call InspectableView(_:parent:index) from CustomControlStyle.inspect(isDragging:) because view.content is not visible.

However, I have observed that this works just as well:

```Swift
extension CustomControlStyle {
	func inspect(isDragging: Bool) throws -> InspectableView<ViewType.ParentView> {
		let configuration = MyControlStyleConfiguration(isDragging: isDragging)
		return makeBody(configuration: configuration).inspect()
	}
}
```

The only difference is it returns an InspectableView<ViewType.ParentView>, rather than an InspectableView<ViewType.ClassifedView). Hence the question concerning the difference between ViewType.ParentView and ViewType.ClassifiedView.

**Reply by gili-labs** [reply]
I was playing with the notion of adding the following extension to InspectableView:

```Swift
@available(iOS 13.0, macOS 10.15, tvOS 13.0, *)
public extension InspectableView {
    func inspectContent() throws -> InspectableView<ViewType.ClassifiedView> {
        return try .init(content, parent: nil, index: nil)
    }
}
```

This would allow me to do something like this:

```Swift
extension DefaultResizableFrameHandleStyle {
    func inspect(
        width: Binding<CGFloat?>,
        height: Binding<CGFloat?>,
        isDragging: Bool) throws -> InspectableView<ViewType.ParentView>
    {
        let configuration = ResizableFrameHandleStyleConfiguration(
            width: width,
            height: height,
            isDragging: isDragging)
        let view = try makeBody(configuration: configuration).inspect()
        return try view.inspectContent()
    }
}
```

This returns an InspectableView with the proper semantics and emulates what you have done already to support SwiftUI styles.

However, I was struggling with the name of the method. I still think this name is alright, but was also toying with the name classify(), which would utterly imply the type of InspectableView returned.

**Reply by nalexn** [reply]
@gili-labs saw this in your PR, good job! I can see why you needed the `classify()`, I managed to refactor the original API so this call won't be needed. You can see the current work version in the [0.7.0](https://github.com/nalexn/ViewInspector/tree/0.7.0) branch.

Also, I saw your other comment about some progress with Gestures - nice! I've been doing some work in parallel with you, so please check out the dev branch I mentioned above - there are some major improvements. Also, I noticed `swiftlint` was missing on my new machine and the code styling went out of control a bit - that's fixed, but I also made it required. Hope that's ok with you

**Reply by gili-labs** [reply]
@nalexn sorry for the delay in responding. I had four glorious days to work on my stuff, and then yesterday it was back to the grind.

I looked at the refactoring you did. This definitely makes the code more straightforward. I also like the other improvements you did around environment variables.

**Reply by nalexn** [reply]
Cool! Let me know if I should hold off the release I plan by the end of this week. It has a very important fix for issues around EnvObjects (#62, #87), I finally found a way to inject EnvObjects without appealing to SwiftUI's native mechanism (which requires hosting of views).

**Reply by gili-labs** [reply]
I still have a lot of testing to do to DragGesture support and I'm probably not going to have time to address it until this weekend. Given the priority on the fixes you've made, I'd go ahead and release as soon as you're ready.
